Documentation/dotnet-primer.md

```diff
@@ -0,0 +1,147 @@
+
+	Dog dog = Dog._nextDogToBeAdopted; // will throw - this is a private field
+
+Generic Types
+-------------
+
+Generic types enable a kind of uncoordinated mix-ins of (largely) arbitrary types. They allow the creation of a kind of _compound type_. A generic type is considered to be _open_, as it requires a set of types (however many type parameters it exposes) to be provided in order to be _closed_. While static in nature, like the rest of the .NET type system, generics provide a kind of faux dynamism that developers find very useful.
+
+The most commonly used generic type is `List<T>` and its sibling `Dictionary<K,V>`. You can use these types, and also create your own generic types. The following example demonstrates the use of these two built-in generic types.
+
+	List<string> strings = new List<string>();
+	strings.Add("string A");
+	strings.Add("string B");
+
+	Dictionary<int,List<string>> lists = new Dictionary<int,List<string>>();
+	lists.Add(DateTime.Now.Hour,strings); // new strings added every hour
+
+Async Programming
+-----------------
+
+Async is a first-class concept within .NET, with async support in the runtime, the framework libraries and various .NET languages. Async is based off of the `Task` concept, which encaptsulates a set of operations to be completed. Tasks are distinct from threads and may not rely on threads or require CPU time much at all, particularly for i/o-bound tasks.
+
+C# includes special treatment for async, including the special keyword `await` for managing tasks. The following example demonstrates calling a web service as an async operation.
+
+	string url = "http://someUrl";
+	HttpClient client = new HttpClient();
+	string json = await client.GetStringAsync(url);
+
+The call to `client.GetStringAsync(url)` does not block, but instead immediately yields by returing a `Task`. Computation resumes and the call returns the requested string when the network activity has completed.
+
+Language Integrated Query (Linq)
+--------------------------------
+
+.NET programs typically operate on some form of data. The data can be database-resident or in the form of objects (sometimes called POCOs - "Plain Old CLR Objects"). Linq provides a language-integrated uniform query model over data, independent of the source. Linq providers bridge the gap between the uniform query model and the form of the data, such as SQL Server tables.
+
+The follow examples demonstrate various uses of Linq to query different forms of data.
+
+TODO: Examples.
+
+Lambdas
+-------
+
+Lambdas are a kind of mini-method. They declare a signature and a method body, but don't have an identity of their own, like a method on a type. They can however, be assigned to a variable and be called, with the appropriate arguments, and passed in a method signature. They are effectively an alternative delegate syntax. Unlike delegates, they can be directly assigned as the left-hand side of event registration or as a Linq select clause.
```

Comment:
Delegates haven't been discussed yet, might be good to do both in this section (i.e. calling it "Delegates and Lambdas")